accepted/2024/[TS]PMUI-Readme-rendering.md

```diff
@@ -0,0 +1,88 @@
+
+We can use the Preview builder as follows:
+```C#
+//This creates a new instance of the preview builder
+var markdownPreview = new PreviewBuilder().Build();
+
+//We update the current markdown being rendered by calling "UpdateContentAsync"
+markdownPreview.UpdateContentAsync(markDown ?? string.Empty, ScrollHint.None)
+
+//IMarkdownPreview.VisualElement contains the FrameworkElement to be passed to the view
+MarkdownPreviewControl = markdownPreview.VisualElement
+```
+
+For packages that are already on the disk we will retrieve the ReadMe from the root folder of the package. We will need to confirm if the current API returns the ReadMe when fetching package details and determine which API we should use to get the data.
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+
+## Drawbacks
+
+This control is not currently using Daytona. There are plans to upgrade it, [Experience 1870733](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1870733), but they have been delayed. 
+
+It's also currently marked as obsolete since the interface has not been finalized. So when an upgrade is made we may have to change how we use the control.
+
+Currently WebView2 controls always render ontop of other controls in the view. [Secnario 25254665](https://microsoft.visualstudio.com/Edge/_workitems/edit/25254665). PM UI needs to be updated to ensure items don't scroll off screen.
+
+## Rationale and alternatives
+By using an existing control we maintain consistency throughout the IDE and can rely on the owner to fix any bugs with the control.
+<!-- Why is this the best design compared to other designs? -->
+<!-- What other designs have been considered and why weren't they chosen? -->
+<!-- What is the impact of not doing this? -->
+
+## Prior Art
+The IMarkdownPreview is currently being used when creating a new pull request inside of Visual Studio.  
+![Alt text](../../meta/resources/ReadMePMUI/PullRequestExperience.png) 
+<!-- What prior art, both good and bad are related to this proposal? -->
+<!-- Do other features exist in other ecosystems and what experience have their community had? -->
+<!-- What lessons from other communities can we learn from? -->
+<!-- Are there any resources that are relevant to this proposal? -->
+
+## Unresolved Questions
+1. Will we show readme in the updates tab?
+1. What do we show if the package has a ReadMe.txt instead of md?
+    * Current suggestion is to focus on ReadMe.md. Once that's complete see how ReadMe.txt would appear when rendered. 
+    * How many packages have a ReadMe.txt vs ReadMe.md?
+1. What do we show if there is no ReadMe defined?
+    - [ ] Hide the tab from the the details pane
+    - [ ] Display the tab with a message saying there is no ReadMe defined.
+1. Where do we get the ReadMe from when it's not on the disk?
```

Comment:
@NuGet/nuget-client from looking through the API docs it doesn't look like there's a specific end point for downloading the ReadMe files. Is there any concern with downloading the nupkg to extract the readme from it when trying to render the ReadMe? This would mean that every time a user looks at a package we would download the file even though they haven't hit install.

Reply:
Does the server "host" these files somewhere similar to the icons, license files etc?

That's the approach we take for license/icons.

There are concerns with download packages we aren't going to use, in particular due to https://learn.microsoft.com/en-us/nuget/concepts/package-installation-process, it'd basically mess with package source mapping.

Reply:
I just posted to the server channel to see if they have an API for the ReadMe. I didn't see anything documented though. The [ReadMeUrl] (https://learn.microsoft.com/en-us/nuget/api/registration-base-url-resource#catalog-entry) in the package metadata is a link to ReadMe section in nuget.org. 

Also for private feeds it's possible that these endpoints aren't implemented fully, for example you search for newtonsoft.json the ReadMe link is missing in our private feeds, but it's there if you use nuget.org as a source. So I guess another implementation detail would be if should we rely on the private feed to source the readme?

Reply:
Where is no endpoint for getting the ReadMe so we're going to focus on rendering the ReadMe for downloaded packages only.